### PR TITLE
ghc: Add `enableDwarf` flag, off by default.

### PR DESCRIPTION
**Pull request by nh2**
This allows our GHCs to build programs with DWARF debug information when `-g` is passed, see https://ghc.haskell.org/trac/ghc/wiki/DWARF.

~On by default on non-Darwin only because `elfutils` (which GHC uses for this) works only for ELF executables.~ Off by default until [GHC ticket #15960 - Using -g causes differences in generated core](https://ghc.haskell.org/trac/ghc/ticket/15960) is fixed.

PR is against `staging` so that the impact can be properly evaluated.

###### Motivation for this change

This can make debugging of many Haskell bugs and performance issues significantly easier.

Common use cases are:

* load the core dumps in gdb and see from what Haskell segfaulting C code is called
* set breakpoints and use valgrind to debug malloc issues
* you can also use `perf` and `kcachegrind` to debug performance issues but results here are mixed

~This change is expected to increase the size of GHC and Haskell libraries, as debugging symbols are added to the object files.

The size of the eventual Haskell binaries that are not explicitly built with `ghc -g` should remain unaffected when the standard statish linking is used (that links Haskell libraries statically, and system libraries dynamically, which is nixpkgs' default).

Closure size should rise minimally to the addition of `elfutils` that provides `libdw`.

I believe that the slightly increased download size for developers will be tremendously offset by the benefits of being able to trouble-shoot crashes and performance issues without having to recompile the entire ecosystem.
(This is from my own perspective of using Haskell for work every day for many years, but I'm relatively sure that others feel the same.)

CC @bgamari @angerman @Gabriel439 @vaibhavsagar @basvandijk @roelvandijk 

###### Things done

Much of the below isn't ticked yet because building takes a while and I added it for all GHC versions.

I have tested the change on 8.2.2 on top of `release-18.03` so far, where I managed to produce dwarf symbols.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

## TODO list

* [ ] Leave some time to collect info on who's onboard with this change
* [x] Fix [forgotten `GhcLibHcOpts = -g3` and `GhcRtsHcOpts = -g3` entries](https://github.com/bgamari/ghc-utils/blob/9f5a7dc633e567f193027685174736c83ed2520a/rel-eng/bin-release.sh#L182-L188) (same as, from https://github.com/NixOS/nixpkgs/pull/52255#issuecomment-447597748)
  * [x] Fix [build error](https://github.com/NixOS/nixpkgs/pull/52255#issuecomment-447610795)
* [x] Figure out how to make it built by Hydra even though it's off by default for now
  * @basvandijk showed me how to do it, adding an attribute set [here](https://github.com/NixOS/nixpkgs/blob/47a4d99c4feae07ca5c9cbaa4888e4081c20cdbb/pkgs/top-level/haskell-packages.nix#L123) which will automatically be built on Hydra by [`release.nix`](https://github.com/NixOS/nixpkgs/blob/47a4d99c4feae07ca5c9cbaa4888e4081c20cdbb/pkgs/top-level/release.nix#L180) (see e.g. how the `integer-simple` one is built [here](https://hydra.nixos.org/job/nixpkgs/trunk/haskell.compiler.integer-simple.ghc822.x86_64-linux)/[here](https://hydra.nixos.org/build/85586694))
* [ ] Do some testing if everything works as expected

### CODE REVIEW

**Comment by domenkozar** [comment]
+1

**Comment by cartazio** [comment]
this wont work on darwin fyi :(

File: pkgs/development/compilers/ghc/8.2.2.nix [hunk 1]
```diff
@@ -17,6 +18,9 @@
   # library instead of the faster but GPLed integer-gmp library.
   enableIntegerSimple ? !(stdenv.lib.any (stdenv.lib.meta.platformMatch stdenv.hostPlatform) gmp.meta.platforms), gmp
 
+, # Allows this GHC to produce binaries that have DWARF debug symbols.
+  enableDwarf ? (!stdenv.isDarwin) # elfutils libdw supports only ELF
```
Comment (nh2) [inline pkgs/development/compilers/ghc/8.2.2.nix]: Is there a better expression than `!stdenv.isDarwin` I could use that gives me "all ELF platforms"?

**Comment by nh2** [comment]
> this wont work on darwin fyi :(

Yes just did a force push to correct that and updated the PR description.

**Comment by cartazio** [comment]
additionally, theres some sort limit that building base with -g tickles in the macho format (at least when i tried to )

**Comment by bgamari** [comment]
It seems to me that this would go nicely with the DWARF-enabled package set I proposed in #51987.

**Comment by nh2** [comment]
@matthewbauer 

> Are ghc binaries built with debugging information by default?

Do you mean in general, or with this PR?

In general, GHC HQ provides bindists in both a normal and a `-dwarf` flavour.

For this PR:

* `--enable-dwarf-unwind` only enables the stack unwinding functionality (printing of the curent stack trace if the program/libraries are compiled with `-g`), currently triggered by either of
  * The GHC [RTS crashing](https://github.com/ghc/ghc/blob/442be690e1ca162e22d632df695c29f90625f33e/rts/RtsMessages.c#L170) (you get a stack trace when `Please report this as a GHC bug` happens in the RTS)
  * if you [send `SIGQUIT`](https://github.com/ghc/ghc/blob/442be690e1ca162e22d632df695c29f90625f33e/rts/posix/Signals.c#L523) to GHC
* we compile all the base libraries used by ghc (`base`, `bytestring`, `vector` etc.) with `-g`, as well as the RTS C code, same as shown [here](https://github.com/bgamari/ghc-utils/blob/9f5a7dc633e567f193027685174736c83ed2520a/rel-eng/bin-release.sh#L182-L188)
  * [ ] but I forgot to put that in (it isn't enabled by by `--enable-dwarf-unwind`), I'll fix that in a moment
* when the above two are done, the final GHC binary itself will have debugging information in, it is not stripped, as the linker that builds it will simply copy all the symbols from all the `.o` files into the binary
  * so if the nix `installPhase` just calls `make install` instead of `make install-strip` (which I think is the case), the GHC binary itself will show a nice stack trace with all info when it crashes _in the RTS_ (as of writing, for GHC itself, only the RTS panic function `rtsFatalInternalErrorFn()` prints a stack trace, Haskell panics don't)
* when GHC is built with `--enable-dwarf-unwind`, you can use the [GHC.ExecutionStack](https://hackage.haskell.org/package/base-4.12.0.0/docs/GHC-ExecutionStack.html) module in your built programs, in particular the `getStackTrace`/`showStackTrace` functions
* An important [note](https://ghc.haskell.org/trac/ghc/wiki/DWARF#OpenIssues) is:
  > While this sort of debugging does not *require* all libraries and RTS to be built with `-g`, it certainly is a very good idea. After all, currently stack unwinding stops at the first closure that doesn't have DWARF information associated with it.
  This is the reason why we want to have `-g` in GHC's base libraries, because GHC's base libraries (e.g. the version of `bytestring` provided by GHC) are used unless you explicitly add a different version to your GHC package database on top

> How does this interact with cabal's -O1 setting?

Cabal may decide to not pass `-g` to GHC when building things, or to strip away debug symbols at the end of the build.

That is controlled by the `cabal` / `Setup.hs --configure` CLI flags:

```
--enable-executable-stripping / disable-executable-stripping
--enable-library-stripping / disable-library-stripping
--enable-debug-info[=n] / --disable-debug-info
```

This PR is only about providing GHC and its base libraries with `-g` and enabling the stack printing functionality; the effort to convince Cabal to produce `haskellPackages` package sets with `-g` is in the other PR https://github.com/NixOS/nixpkgs/issues/51987#issuecomment-447214739 (which needs this PR).

File: pkgs/development/compilers/ghc/8.2.2.nix (hunk 1 above)
Comment (nh2) [reply pkgs/development/compilers/ghc/8.2.2.nix]: @angerman on `#ghc` chat said:

> I doubt there is any consumer of Mach-o that is not macOS or iOS or really exotic. Same for Windows. Everyone else just uses elf.

**Comment by nh2** [comment]
After having [added](https://github.com/NixOS/nixpkgs/pull/52255/commits/fa39fa4033841539601786458c399a1fca74b2e7#diff-c90b207a56851dba68c66cf16a911e8eR65) the forgotten `GhcLibHcOpts = -g3` and `GhcRtsHcOpts = -g3` entries, I get this error when building GHC:

```
/build/ghc1518_0/ghc_4.s: Assembler messages:

/build/ghc1518_0/ghc_4.s:1719:0: error:
     Error: invalid operands (.debug_frame and .note.GNU-stack sections) for `-'
     |
1719 |         .uleb128 1f-.-1
     | ^
```

**Comment by nh2** [comment]
> I get this error

@bgamari said:

> IIRC it's a binutils bug
> but we worked around it in 8.4.3
> indeed it looks like it's fixed upstream: https://sourceware.org/bugzilla/show_bug.cgi?id=23040
> but the fix hasn't made it into a release yet

nh2:

> ah. Does "worked around" mean "in the GHC tree"? If yes, I could just skip building dwarf for 8.2.2

@bgamari:

> yes

**Comment by nh2** [comment]
Hmm, I got the assembler error even when building `/nix/store/62qbyv9d40kzaz6mn9v28q5l7m35gng5-ghc-8.5.20180118.drv`.

It looks like I only get it on `ghc/head.nix` which is currently at `8.5.20180118` - surprisingly old https://github.com/NixOS/nixpkgs/blob/3dbf0f9f58a672dee62a6ea776d58f8057029627/pkgs/development/compilers/ghc/head.nix#L29. Probably too old.

8.4.4 and the 8.6.* releases all built.

**Comment by nh2** [comment]
Here are some numbers for binary sizes:

For the Haskell `hello` application built on this branch with `NIX_PATH=nixpkgs=. nix-build -E 'with (import <nixpkgs> {}); haskell.packages.dwarf.ghc844.hello'`.

I've checked the presence of debugging symbols using `objdump -Wi /nix/store/x1l9gjvk5jhdn9rf6cvbryilqz22ij7l-hello-1.0.0.2/bin/hello`.

The program has "statish" linking:

<details>

```
% ldd /nix/store/x1l9gjvk5jhdn9rf6cvbryilqz22ij7l-hello-1.0.0.2/bin/hello
	linux-vdso.so.1 =>  (0x00007fff9713f000)
	libm.so.6 => /nix/store/fa2sgggx44vak9fsw3zgzv8xbh99a8hc-glibc-2.27/lib/libm.so.6 (0x00007fe100ffe000)
	libgmp.so.10 => /nix/store/f5wq2jnpi80pax6n63asyj095f2c1lp1-gmp-6.1.2/lib/libgmp.so.10 (0x00007fe100d6a000)
	librt.so.1 => /nix/store/fa2sgggx44vak9fsw3zgzv8xbh99a8hc-glibc-2.27/lib/librt.so.1 (0x00007fe100b62000)
	libdl.so.2 => /nix/store/fa2sgggx44vak9fsw3zgzv8xbh99a8hc-glibc-2.27/lib/libdl.so.2 (0x00007fe10095e000)
	libelf.so.1 => /nix/store/jjmdb0cg43l0b1abpphgi4ikxvych6hc-elfutils-0.175/lib/libelf.so.1 (0x00007fe100745000)
	libdw.so.1 => /nix/store/jjmdb0cg43l0b1abpphgi4ikxvych6hc-elfutils-0.175/lib/libdw.so.1 (0x00007fe1004f3000)
	libc.so.6 => /nix/store/fa2sgggx44vak9fsw3zgzv8xbh99a8hc-glibc-2.27/lib/libc.so.6 (0x00007fe10013f000)
	libpthread.so.0 => /nix/store/fa2sgggx44vak9fsw3zgzv8xbh99a8hc-glibc-2.27/lib/libpthread.so.0 (0x00007fe0fff20000)
	/nix/store/fa2sgggx44vak9fsw3zgzv8xbh99a8hc-glibc-2.27/lib/ld-linux-x86-64.so.2 => /lib64/ld-linux-x86-64.so.2 (0x00007fe101393000)
	libz.so.1 => /nix/store/hplrkgq6qikida9zjbnlm5xk4r35d53w-zlib-1.2.11/lib/libz.so.1 (0x00007fe0ffd04000)
	liblzma.so.5 => /nix/store/xkn2jyrs31ddiyv04lypg814qqq8pmw2-xz-5.2.4/lib/liblzma.so.5 (0x00007fe0ffadd000)
	libbz2.so.1 => /nix/store/7pqz1wamgajar7c61g1dsg5475bjfvf8-bzip2-1.0.6.0.1/lib/libbz2.so.1 (0x00007fe0ff8cd000)
```

</details>

Sizes (normal, with `strip` run on it, and `strip -S`, which strips debug symbols only, run on it; each compressed with `gzip` and `xz` default params on Ubuntu 16.04):

```
hello                   985 KB
hello-stripped-S        970 KB
hello-stripped          781 KB

hello.gz                286 KB
hello-stripped-S.gz     280 KB
hello-stripped.gz       235 KB

hello.xz                223 KB
hello-stripped-S.xz     219 KB
hello-stripped.xz       184 KB
```

**Comment by nh2** [comment]
@GrahamcOfBorg build haskell.packages.dwarf.ghc863.hello

**Comment by nh2** [comment]
Why does `grahamcofborg-eval` say

```
evaluation aborted with the following error message: 'cannot find attribute `haskell.compiler.dwarf.ghc844''
```

? I can `nix-instantiate` that on my machine on this PR.

**Comment by nh2** [comment]
More statistics on a larger package with many dependencies:

On this PR, `stack` via `NIX_PATH=nixpkgs=. nix-build -E 'with (import <nixpkgs> {}); haskell.packages.dwarf.ghc863.stack'`:

```
stack                139M
stack-stripped-S      40M
stack-stripped        24M

stack.gz              22M
stack-stripped-S.gz  9.0M
stack-stripped.gz    5.9M

stack.xz              14M
stack-stripped.xz    3.6M
stack-stripped-S.xz  5.5M
```

So here, the DWARF overhead is **3.5x** over nixpkgs' normal builds (and 3x when xz-compressed).

**Comment by nh2** [comment]
ofborg seems to be unhappy with me:

* https://gist.github.com/GrahamcOfBorg/84a2d291d9d35fa5ab6a32ee70bf1cb2
  ```
  cannot find attribute `haskell.compiler.dwarf.ghc844'
  ```
  no idea why
* ARM build https://github.com/NixOS/nixpkgs/pull/52255/checks?check_run_id=41635996
  * `No space left on device`
* x86_64-linux:
  * `timed out after 1800 seconds`

**Comment by Infinisil** [comment]
Btw, if you wanna be pragmatic about the elf thing, you could use `stdenv.hostPlatform.parsed.kernel.execFormat == stdenv.lib.systems.parse.execFormats.elf`

(I'm not sure if `hostPlatform` or `targetPlatform` or `buildPlatform` is wanted here though)

**Comment by Ericson2314** [comment]
We can add an `isElf` and `isMachO` to https://github.com/NixOS/nixpkgs/blob/master/lib/systems/inspect.nix

**Comment by Ericson2314** [comment]
Overall, really nice PR @nh2. The comments / small refactors to existing stuff too really shows the level of care :).

**Comment by nh2** [comment]
> (I'm not sure if `hostPlatform` or `targetPlatform` or `buildPlatform` is wanted here though)

@angerman Do you know that?

**Comment by nh2** [comment]
> https://git.haskell.org/ghc.git/commitdiff_plain/cb882fc993b4972f7f212b291229ef9e9ade0af9 can we just fetchPatch that instead?

@Ericson2314 No unfortunately, because that commit actually has a bug: It lacks a comma in the last `+` line. (Obviously checks for syntactical correctness are a foreign concept to `autoconf`, so it went undetected for a while, and was fixed up in [this commit](https://gitlab.haskell.org/ghc/ghc/commit/126aa794743b807b7447545697b96dd165ec3e16).)

We could `fetchpatch` both commits, but at the time of writing the second one wasn't in GHC yet, and by now I'm still not sure if applying a wrong patch and then fixing it up with another one would be easier to inspect than just reading one syntax-correct patch.

**Comment by Ericson2314** [comment]
@nh2 Ah ok. That's fine I suppose. Maybe just add a comment so the next person knows everything has been upstreamed in some form?

**Comment by nh2** [comment]
> Maybe just add a comment so the next person knows everything has been upstreamed in some form?

Good point, done.

**Comment by nh2** [comment]
I've also added a commit that uses `--enable-debug-info` instead of `--ghc-options=-g`

**Review by bgamari — APPROVED** [review]
I'm looking forward to having this. Thanks @nh2!

**Comment by bgamari** [comment]
That being said, I wonder if a note in the users guide is in order? Otherwise I'm a bit worried that the new package set will be quite undiscoverable.

File: pkgs/top-level/haskell-packages.nix [hunk 2]
```diff
@@ -174,18 +209,39 @@ in {
...
         integer-simple = null;
+        # Get rid of `integer-gmp` as a safety measure to notice
+        # when anything hard-depends on it.
         integer-gmp = null;
       };
     });
 
+    # The dwarf attribute set contains package sets for all the GHC compilers
+    # with all packages compiled with `-g` debugging info.
+    dwarf = pkgs.lib.genAttrs normalGhcCompilers (name: packages."${name}".override {
```
Comment (bgamari) [inline pkgs/top-level/haskell-packages.nix]: @nh2, I just realized that with the current construction it is not possible to add overrides to the `dwarf` package set. This seems like a major limitation that I will inevitably run into.

Comment (nh2) [reply pkgs/top-level/haskell-packages.nix]: Hmm, anybody have an idea what can be done about this?

Wouldn't that problem exist for all other package sets as well?

**Review by bgamari — CHANGES_REQUESTED** [review]
I think we should work out the overriding story before we commit to this since the current situation is unfortunately entirely non-composable. Perhaps instead of adding the `dwarf` package sets here we could just expose something like,
```nix
let withDwarfEnabled = name: args: packages."${name}".override (args // {
  ghc = bh.compiler.dwarf."${name}";
  buildHaskellPackages = bh.packages.dwarf."${name}";
  overrides =
    let
      dwarfOverrides =
        self : super : {
        # Override the mkDerivation function so that the GHC flags that
        # are needed for debugging symbols are passed to all Cabal invocations.
        mkDerivation = old: super.mkDerivation (old // {
          configureFlags = (old.configureFlags or []) ++ [
            "--enable-debug-info=3"
            "--disable-library-stripping"
            "--disable-executable-stripping"
          ];
          # Disable nixpkgs' own stripping.
          dontStrip = true;
        });
      };
    in composeExtensions dwarfOverrides args.overrides;
});
```
This significantly less elegant and less convenient than the current package sets but is nevertheless sufficient and avoids committing us to an approach which is fundamentally inextensible.

Perhaps a more elegant (albeit still not ideomatic) approach would be to expose an attribute set of partial applications of `withDwarfEnabled` as defined above:
```
dwarf = pkgs.lib.genAttrs normalGhcCompilers withDwarfEnabled
```
The user would then have to apply an empty overrides set to an attribute of `dwarf` to get a Haskell packages set.

File: pkgs/top-level/haskell-packages.nix (hunk 2 above)
Comment (bgamari) [reply pkgs/top-level/haskell-packages.nix]: Indeed, I believe the `integer-simple` set is already broken in this same way.

**Comment by domenkozar** [comment]
@nh2 any chance to rebase this? I'd love to have it in 19.03 - even if overriding is broken for now.

**Comment by nh2** [comment]
@domenkozar I agree, even if the package set isn't super useful, it's good to be able to have at least the pre-built compiler downloadable from cache.nixos.org, and the ability to build many Haskell executables with DWARF.

I just pushed https://github.com/nh2/nixpkgs/tree/ghc-dwarf-19.03 but it's still building, so not sure yet if it works.

**Comment by matthewbauer** [comment]
Why not just use enableSeparateDebugInfo for all compilers? That way we don’t have to build every compiler 2 times. Those patches should be okay even when you don’t want dwarf symbols.

**Comment by nh2** [comment]
> I just pushed https://github.com/nh2/nixpkgs/tree/ghc-dwarf-19.03 but it's still building, so not sure yet if it works.

I've built `stack` with it now, so the rebase worked. I also pushed it to the `static-haskell-nix` cachix (commit d9bda1fed6ed5041242ea9e7917583a255c4e9b4, build command was `nix-build $(NIX_PATH=nixpkgs=. nix-instantiate -E 'with (import <nixpkgs> {}); haskell.packages.dwarf.ghc863.stack')`).

**Comment by nh2** [comment]
> Why not just use enableSeparateDebugInfo for all compilers? That way we don’t have to build every compiler 2 times. Those patches should be okay even when you don’t want dwarf symbols.

@matthewbauer Can you elaborate a bit? I'm not super familiar with how `separateDebugInfo` works, and if it can easily be used with ghc.

Also, what do you mean with "those patches"?

**Comment by matthewbauer** [comment]
My main concern is this doubles the number of GHC compilers we build. I like enabling debug info, but that should be in a separate debug output not a whole new compiler. ```separateDebugInfo``` strips out all of the debug symbols and puts the symbols in the ```debug``` identified by the build id. I'd really like to do this everywhere (#18530) but there are concerns with how expensive it would become.

I would propose doing this:

- Set ```enableSeparateDebugInfo = true``` for all ghc compilers along with the ```-g``` flags and any other patches needed to get ghc to include debugging symbols.
- Maybe set ```--enable-dwarf-unwind``` by default for compilers on Linux as long as it doesn't lead to issues in output. It definitely makes sense to disable it on some compilers where it doesn't work.
- Add enableSeparateDebugInfo to the enableDebugging function. Possibly enable debugging for all haskellPackages, if it's not too expensive. /cc @vcunat @edolstra 

As a side note, I also really dislike having integer-simple as a separate package set. It has many of the same problems as a "dwarf" package set, but there's not a good alternative available because GHC requires you to choose an integer library at build time .

**Comment by vcunat** [comment]
We could estimate price for our haskellPackages specifically, if you produce code.  See: https://github.com/NixOS/nixpkgs/issues/18530#issuecomment-469053062

**Comment by nh2** [comment]
@vcunat I left a comment with some basic statistics in https://github.com/NixOS/nixpkgs/issues/18530#issuecomment-469059983

**Comment by vcunat** [comment]
I admit I've read almost nothing from this thread so far :-/  but I tried this PR with `? true` everywhere, and ghc won't even compile.  I'm getting `file number 1 already allocated`; perhaps you know more, as I noticed [this thread](http://haskell.1045720.n5.nabble.com/GHC-15955-Cannot-get-debugging-symbols-for-compiled-c-sources-td5886714.html).  Maybe I'm doing something wrong, though...

**Comment by nh2** [comment]
@vcunat https://github.com/NixOS/nixpkgs/pull/52255#issuecomment-468485577 shows a nix invocation that should work on the branch here in the PR, which builds one large Haskell executable and its dependency closure (which may be interesting to compare with the dependency closure of the one you get if you leave the `.dwarf` away).

I can also offer to collect some info for you, if you let me know what it shall be :)

**Comment by vcunat** [comment]
That evaluates to the same ghc which fails for me, twice now.  That is, after merging this PR with my branch discussed on the other thread: 0173520e0b4.

**Comment by vcunat** [comment]
I can confirm on pure commit 58990361 it works for me.

**Comment by nh2** [comment]
Relevant for GHC 8.10: https://github.com/NixOS/nixpkgs/pull/69552

**Comment by cdepillabout** [comment]
I believe there has been other PRs adding dwarf support:

- https://github.com/NixOS/nixpkgs/pull/68362
- https://github.com/NixOS/nixpkgs/pull/69032
- https://github.com/NixOS/nixpkgs/pull/69552

**Pull request closed without merging**